### PR TITLE
feat(bug-finding): warn users if resource config conflicts with machi…

### LINKED ISSUE #49: feat(bug-finding): warn users if resource config conflicts with actual machine resources
Our default configs at `example_configs/*/config-resource.yaml` typically specify 16 cores and 64GB of memory. Docker will emit an error if such resources aren't available from the user's machine, but we would like to catch this earlier and emit a more informative error message. Something like the following:

> Machine does not have adequate resources! Only <M> CPU cores and <N> memory is available. Please edit <PATH_TO_CONFIG>

### PR DESCRIPTION
**Pull request by tkqdldk**
Closes #49 

## Description

Adds resources check and a warning when the compose file resource configurations (max CPU and total memory) exceed machine resources. The check runs in "Validate Configuration for Running" phase, before Docker attempts to start containers.

The function emits a warning but does not stop the execution.

## Changes
- Added `get_host_total_memory()` in `utils.py`
- Added `_validate_machine_resources()` in `crs_compose.py`
- Added validation task in `__validate_before_run`
- Added changelog entry

## Validation

All existing tests pass.

P.S. : Can't assign a reviewer directly, cc @azchin for review

### CODE REVIEW

File: oss_crs/src/utils.py [hunk 1]
```diff
@@ -165,6 +166,24 @@ def rm_with_docker(path: Path) -> None:
...
+    try:
+        pagesize = os.sysconf("SC_PAGE_SIZE")
+        pages = os.sysconf("SC_PHYS_PAGES")
+        if pagesize > 0 and pages > 0:
+            return pagesize * pages
+
+    except(OSError):
+            pass
+
+    memeinfo = Path("/proc/meminfo")
```
Comment (azchin) [inline oss_crs/src/utils.py]: `meminfo` typo (caught by ruff)

File: oss_crs/src/crs_compose.py [hunk 2]
```diff
@@ -640,6 +644,52 @@ def _validate_required_inputs(
             return TaskResult(success=False, error="\n".join(errors))
         return TaskResult(success=True)
 
+    def _validate_machine_resources(self) -> TaskResult:
+        """Validate that machine resources and resource config do not conflict."""
+        # Get machine CPU count and memory (in Bytes)
+        machine_cpu_count = os.cpu_count()
+        machine_memory = get_host_memory()
+
+        if machine_cpu_count is None or machine_memory == 0:
```
Comment (azchin) [inline oss_crs/src/crs_compose.py]: (C1) Let's get `machine_memory` to return `None` on failure

File: oss_crs/src/utils.py [hunk 3]
```diff
@@ -165,6 +166,24 @@ def rm_with_docker(path: Path) -> None:
     except subprocess.CalledProcessError as e:
         raise RuntimeError(f"Error removing {path} with Docker: {e}")
 
+def get_host_memory() -> int:
```
Comment (azchin) [inline oss_crs/src/utils.py]: Return val as `int | None`, see (C1)

File: oss_crs/src/utils.py [hunk 4]
```diff
@@ -165,6 +166,24 @@ def rm_with_docker(path: Path) -> None:
...
+
+    except(OSError):
+            pass
+
+    memeinfo = Path("/proc/meminfo")
+    if memeinfo.exists():
+      for line in meminfo.read_text().splitlines():
+        if line.startswith("MemTotal:"):
+          return int(line.split()[1])*1024
+    return 0
```
Comment (azchin) [inline oss_crs/src/utils.py]: `return None` see (C1)

File: oss_crs/src/crs_compose.py [hunk 5]
```diff
@@ -640,6 +644,52 @@ def _validate_required_inputs(
...
+            except ValueError as e:
+                log_warning(f"Failed to validate cpuset for {name}: {e}")
+            
+            # Calculate max memory needed
+            try:
+                total_memory_required += parse_memory_to_bytes(cfg.memory)
+            except ValueError as e:
+                log_warning(f"Failed to parse memory for {name}: {e}")
+
+        cpu_check = max_cpus_required <= machine_cpu_count
```
Comment (azchin) [inline oss_crs/src/crs_compose.py]: I think this is an off-by-one, since `os.cpu_count()` returns number of CPUs but cpuset is indexed by zero, so we'd get range of `0-3` on 4-core machine. `max([0, 1, 2, 3]) == 3 < 4`

`max_cpus_required < machine_cpu_count` would be the correct check. But please check/test for this edge case dynamically.

**Review by azchin — CHANGES_REQUESTED** [review]
Thanks for the PR and sorry for the late response! I have a few requests detailed below. We also use a few linters and formatters in our [CI workflow](https://github.com/ossf/oss-crs/blob/ceb67e0813f5d336165fd7ea394572deea59da61/.github/workflows/ci.yml#L28-L35), so I recommend running them manually before pushing.

File: oss_crs/src/crs_compose.py (hunk 5 above)
Comment (tkqdldk) [reply oss_crs/src/crs_compose.py]: You're absolutely right ^^'... I will also add dynamic testing.

**Comment by azchin** [comment]
Hm I'm not seeing the error messages implemented in your PR in my output. The build does indeed abort, but that's because of Docker's error handling and happens after `docker compose` gets run.

I can take a deeper look at what we're missing after the weekend, if needed.

```
% uv run oss-crs build-target --compose-file example/crs-libfuzzer/compose.yaml --fuzz-proj-path ~/post/CRSBench/benchmarks/sanity-mock-c-delta-01
╭─────────────────────────────────────────── Init CRS: crs-libfuzzer ────────────────────────────────────────────╮
│          Task         Status                                                                                   │
│   ✓      Git Fetch    Success                                                                                  │
│   ✓      Git Reset    Success                                                                                  │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭─────────────────────────────────────────── CRS Compose Build Target ───────────────────────────────────────────╮
│          Task                                                           Status                                 │
│   ✗      crs-libfuzzer                                                  Failed                                 │
│            └─ build                                                     Failed                                 │
│              └─ Prepare docker compose file                             Success                                │
│              └─ Prepare docker images defined in docker compose file    Success                                │
│              └─ Build target by executing the docker compose            Failed                                 │
│              └─ Check outputs                                           Pending                                │
│              └─ 🧹 Cleanup Docker Compose                               Success                                │
│ ╭─────────────────────────────────────────── Error: crs-libfuzzer ───────────────────────────────────────────╮ │
│ │ Command failed with exit code 1:                                                                           │ │
│ │ docker compose -p crs_582rjhwn5c -f /tmp/da3y01n0ad/docker-compose.yaml run --rm target_builder            │ │
│ │                                                                                                            │ │
│ │  Network crs_582rjhwn5c_default Creating                                                                   │ │
│ │  Network crs_582rjhwn5c_default Created                                                                    │ │
│ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Creating                                         │ │
│ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Error response from daemon: Requested CPUs are   │ │
│ │ not available - requested 4-129, available: 0-127                                                          │ │
│ │ Error response from daemon: Requested CPUs are not available - requested 4-129, available: 0-127           │ │
│ │ 📝 Docker compose output:                                                                                  │ │
│ │ ---                                                                                                        │ │
│ │ Command failed with exit code 1:                                                                           │ │
│ │ docker compose -p crs_582rjhwn5c -f /tmp/da3y01n0ad/docker-compose.yaml run --rm target_builder            │ │
│ │                                                                                                            │ │
│ │  Network crs_582rjhwn5c_default Creating                                                                   │ │
│ │  Network crs_582rjhwn5c_default Created                                                                    │ │
│ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Creating                                         │ │
│ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Error response from daemon: Requested CPUs are   │ │
│ │ not available - requested 4-129, available: 0-127                                                          │ │
│ │ Error response from daemon: Requested CPUs are not available - requested 4-129, available: 0-127           │ │
│ │ ---                                                                                                        │ │
│ │ 📝 Docker compose file contents:                                                                           │ │
│ │ ---                                                                                                        │ │
│ │ services:                                                                                                  │ │
│ │   target_builder:                                                                                          │ │
│ │     privileged: true                                                                                       │ │
│ │     shm_size: 2g                                                                                           │ │
│ │     environment:                                                                                           │ │
│ │       - HELPER=True                                                                                        │ │
│ │       - RUN_FUZZER_MODE=interactive                                                                        │ │
│ │       - FUZZING_ENGINE=libfuzzer                                                                           │ │
│ │       - SANITIZER=address                                                                                  │ │
│ │       - ARCHITECTURE=x86_64                                                                                │ │
│ │       - PROJECT_NAME=sanity-mock-c-delta-01                                                                │ │
│ │       - FUZZING_LANGUAGE=c                                                                                 │ │
│ │       - OSS_CRS_RUN_ENV_TYPE=local                                                                         │ │
│ │       - OSS_CRS_BUILD_ID=1775158626g9                                                                      │ │
│ │       - OSS_CRS_BUILD_OUT_DIR=/OSS_CRS_BUILD_OUT_DIR                                                       │ │
│ │       - OSS_CRS_TARGET=sanity-mock-c-delta-01                                                              │ │
│ │       - OSS_CRS_PROJ_PATH=/OSS_CRS_PROJ_PATH                                                               │ │
│ │       - OSS_CRS_TARGET_PROJ_DIR=/OSS_CRS_PROJ_PATH                                                         │ │
│ │       - OSS_CRS_REPO_PATH=/src/mock-c                                                                      │ │
│ │     build:                                                                                                 │ │
│ │       context: /home/andrew/post/oss-crs-tkqdldk/.oss-crs-workdir/crs_compose/crs_src/crs-libfuzzer        │ │
│ │       dockerfile:                                                                                          │ │
│ │ /home/andrew/post/oss-crs-tkqdldk/.oss-crs-workdir/crs_compose/crs_src/crs-libfuzzer/oss-crs/dockerfiles/b │ │
│ │ uilder.Dockerfile                                                                                          │ │
│ │       args:                                                                                                │ │
│ │         - target_base_image=sanity-mock-c-delta-01:4c6d77cdc8f0                                            │ │
│ │         - crs_version=1.0.0                                                                                │ │
│ │       additional_contexts:                                                                                 │ │
│ │         libcrs: /home/andrew/post/oss-crs-tkqdldk/libCRS                                                   │ │
│ │     cpuset: "4-129"                                                                                        │ │
│ │     mem_limit: "16G"                                                                                       │ │
│ │     volumes:                                                                                               │ │
│ │       -                                                                                                    │ │
│ │ /home/andrew/post/oss-crs-tkqdldk/.oss-crs-workdir/crs_compose/43e5a06b0d32/address/builds/1775158626g9/cr │ │
│ │ s/crs-libfuzzer/sanity-mock-c-delta-01_4c6d77cdc8f0/BUILD_OUT_DIR:/OSS_CRS_BUILD_OUT_DIR                   │ │
│ │ ---                                                                                                        │ │
│ ╰────────────────────────────────────────────────────────────────────────────────────────────────────────────╯ │
│ ╭─────────────────────────────────────────────── Error: build ───────────────────────────────────────────────╮ │
│ │ Command failed with exit code 1:                                                                           │ │
│ │ docker compose -p crs_582rjhwn5c -f /tmp/da3y01n0ad/docker-compose.yaml run --rm target_builder            │ │
│ │                                                                                                            │ │
│ │  Network crs_582rjhwn5c_default Creating                                                                   │ │
│ │  Network crs_582rjhwn5c_default Created                                                                    │ │
│ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Creating                                         │ │
│ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Error response from daemon: Requested CPUs are   │ │
│ │ not available - requested 4-129, available: 0-127                                                          │ │
│ │ Error response from daemon: Requested CPUs are not available - requested 4-129, available: 0-127           │ │
│ │ 📝 Docker compose output:                                                                                  │ │
│ │ ---                                                                                                        │ │
│ │ Command failed with exit code 1:                                                                           │ │
│ │ docker compose -p crs_582rjhwn5c -f /tmp/da3y01n0ad/docker-compose.yaml run --rm target_builder            │ │
│ │                                                                                                            │ │
│ │  Network crs_582rjhwn5c_default Creating                                                                   │ │
│ │  Network crs_582rjhwn5c_default Created                                                                    │ │
│ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Creating                                         │ │
│ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Error response from daemon: Requested CPUs are   │ │
│ │ not available - requested 4-129, available: 0-127                                                          │ │
│ │ Error response from daemon: Requested CPUs are not available - requested 4-129, available: 0-127           │ │
│ │ ---                                                                                                        │ │
│ │ 📝 Docker compose file contents:                                                                           │ │
│ │ ---                                                                                                        │ │
│ │ services:                                                                                                  │ │
│ │   target_builder:                                                                                          │ │
│ │     privileged: true                                                                                       │ │
│ │     shm_size: 2g                                                                                           │ │
│ │     environment:                                                                                           │ │
│ │       - HELPER=True                                                                                        │ │
│ │       - RUN_FUZZER_MODE=interactive                                                                        │ │
│ │       - FUZZING_ENGINE=libfuzzer                                                                           │ │
│ │       - SANITIZER=address                                                                                  │ │
│ │       - ARCHITECTURE=x86_64                                                                                │ │
│ │       - PROJECT_NAME=sanity-mock-c-delta-01                                                                │ │
│ │       - FUZZING_LANGUAGE=c                                                                                 │ │
│ │       - OSS_CRS_RUN_ENV_TYPE=local                                                                         │ │
│ │       - OSS_CRS_BUILD_ID=1775158626g9                                                                      │ │
│ │       - OSS_CRS_BUILD_OUT_DIR=/OSS_CRS_BUILD_OUT_DIR                                                       │ │
│ │       - OSS_CRS_TARGET=sanity-mock-c-delta-01                                                              │ │
│ │       - OSS_CRS_PROJ_PATH=/OSS_CRS_PROJ_PATH                                                               │ │
│ │       - OSS_CRS_TARGET_PROJ_DIR=/OSS_CRS_PROJ_PATH                                                         │ │
│ │       - OSS_CRS_REPO_PATH=/src/mock-c                                                                      │ │
│ │     build:                                                                                                 │ │
│ │       context: /home/andrew/post/oss-crs-tkqdldk/.oss-crs-workdir/crs_compose/crs_src/crs-libfuzzer        │ │
│ │       dockerfile:                                                                                          │ │
│ │ /home/andrew/post/oss-crs-tkqdldk/.oss-crs-workdir/crs_compose/crs_src/crs-libfuzzer/oss-crs/dockerfiles/b │ │
│ │ uilder.Dockerfile                                                                                          │ │
│ │       args:                                                                                                │ │
│ │         - target_base_image=sanity-mock-c-delta-01:4c6d77cdc8f0                                            │ │
│ │         - crs_version=1.0.0                                                                                │ │
│ │       additional_contexts:                                                                                 │ │
│ │         libcrs: /home/andrew/post/oss-crs-tkqdldk/libCRS                                                   │ │
│ │     cpuset: "4-129"                                                                                        │ │
│ │     mem_limit: "16G"                                                                                       │ │
│ │     volumes:                                                                                               │ │
│ │       -                                                                                                    │ │
│ │ /home/andrew/post/oss-crs-tkqdldk/.oss-crs-workdir/crs_compose/43e5a06b0d32/address/builds/1775158626g9/cr │ │
│ │ s/crs-libfuzzer/sanity-mock-c-delta-01_4c6d77cdc8f0/BUILD_OUT_DIR:/OSS_CRS_BUILD_OUT_DIR                   │ │
│ │ ---                                                                                                        │ │
│ ╰────────────────────────────────────────────────────────────────────────────────────────────────────────────╯ │
│ ╭─────────────────────────── Error: Build target by executing the docker compose ────────────────────────────╮ │
│ │ Command failed with exit code 1:                                                                           │ │
│ │ docker compose -p crs_582rjhwn5c -f /tmp/da3y01n0ad/docker-compose.yaml run --rm target_builder            │ │
│ │                                                                                                            │ │
│ │  Network crs_582rjhwn5c_default Creating                                                                   │ │
│ │  Network crs_582rjhwn5c_default Created                                                                    │ │
│ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Creating                                         │ │
│ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Error response from daemon: Requested CPUs are   │ │
│ │ not available - requested 4-129, available: 0-127                                                          │ │
│ │ Error response from daemon: Requested CPUs are not available - requested 4-129, available: 0-127           │ │
│ ╰────────────────────────────────────────────────────────────────────────────────────────────────────────────╯ │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

**Comment by tkqdldk** [comment]
> Hm I'm not seeing the error messages implemented in your PR in my output. The build does indeed abort, but that's because of Docker's error handling and happens after `docker compose` gets run.
> 
> I can take a deeper look at what we're missing after the weekend, if needed.
> 
> ```
> % uv run oss-crs build-target --compose-file example/crs-libfuzzer/compose.yaml --fuzz-proj-path ~/post/CRSBench/benchmarks/sanity-mock-c-delta-01
> ╭─────────────────────────────────────────── Init CRS: crs-libfuzzer ────────────────────────────────────────────╮
> │          Task         Status                                                                                   │
> │   ✓      Git Fetch    Success                                                                                  │
> │   ✓      Git Reset    Success                                                                                  │
> ╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
> ╭─────────────────────────────────────────── CRS Compose Build Target ───────────────────────────────────────────╮
> │          Task                                                           Status                                 │
> │   ✗      crs-libfuzzer                                                  Failed                                 │
> │            └─ build                                                     Failed                                 │
> │              └─ Prepare docker compose file                             Success                                │
> │              └─ Prepare docker images defined in docker compose file    Success                                │
> │              └─ Build target by executing the docker compose            Failed                                 │
> │              └─ Check outputs                                           Pending                                │
> │              └─ 🧹 Cleanup Docker Compose                               Success                                │
> │ ╭─────────────────────────────────────────── Error: crs-libfuzzer ───────────────────────────────────────────╮ │
> │ │ Command failed with exit code 1:                                                                           │ │
> │ │ docker compose -p crs_582rjhwn5c -f /tmp/da3y01n0ad/docker-compose.yaml run --rm target_builder            │ │
> │ │                                                                                                            │ │
> │ │  Network crs_582rjhwn5c_default Creating                                                                   │ │
> │ │  Network crs_582rjhwn5c_default Created                                                                    │ │
> │ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Creating                                         │ │
> │ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Error response from daemon: Requested CPUs are   │ │
> │ │ not available - requested 4-129, available: 0-127                                                          │ │
> │ │ Error response from daemon: Requested CPUs are not available - requested 4-129, available: 0-127           │ │
> │ │ 📝 Docker compose output:                                                                                  │ │
> │ │ ---                                                                                                        │ │
> │ │ Command failed with exit code 1:                                                                           │ │
> │ │ docker compose -p crs_582rjhwn5c -f /tmp/da3y01n0ad/docker-compose.yaml run --rm target_builder            │ │
> │ │                                                                                                            │ │
> │ │  Network crs_582rjhwn5c_default Creating                                                                   │ │
> │ │  Network crs_582rjhwn5c_default Created                                                                    │ │
> │ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Creating                                         │ │
> │ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Error response from daemon: Requested CPUs are   │ │
> │ │ not available - requested 4-129, available: 0-127                                                          │ │
> │ │ Error response from daemon: Requested CPUs are not available - requested 4-129, available: 0-127           │ │
> │ │ ---                                                                                                        │ │
> │ │ 📝 Docker compose file contents:                                                                           │ │
> │ │ ---                                                                                                        │ │
> │ │ services:                                                                                                  │ │
> │ │   target_builder:                                                                                          │ │
> │ │     privileged: true                                                                                       │ │
> │ │     shm_size: 2g                                                                                           │ │
> │ │     environment:                                                                                           │ │
> │ │       - HELPER=True                                                                                        │ │
> │ │       - RUN_FUZZER_MODE=interactive                                                                        │ │
> │ │       - FUZZING_ENGINE=libfuzzer                                                                           │ │
> │ │       - SANITIZER=address                                                                                  │ │
> │ │       - ARCHITECTURE=x86_64                                                                                │ │
> │ │       - PROJECT_NAME=sanity-mock-c-delta-01                                                                │ │
> │ │       - FUZZING_LANGUAGE=c                                                                                 │ │
> │ │       - OSS_CRS_RUN_ENV_TYPE=local                                                                         │ │
> │ │       - OSS_CRS_BUILD_ID=1775158626g9                                                                      │ │
> │ │       - OSS_CRS_BUILD_OUT_DIR=/OSS_CRS_BUILD_OUT_DIR                                                       │ │
> │ │       - OSS_CRS_TARGET=sanity-mock-c-delta-01                                                              │ │
> │ │       - OSS_CRS_PROJ_PATH=/OSS_CRS_PROJ_PATH                                                               │ │
> │ │       - OSS_CRS_TARGET_PROJ_DIR=/OSS_CRS_PROJ_PATH                                                         │ │
> │ │       - OSS_CRS_REPO_PATH=/src/mock-c                                                                      │ │
> │ │     build:                                                                                                 │ │
> │ │       context: /home/andrew/post/oss-crs-tkqdldk/.oss-crs-workdir/crs_compose/crs_src/crs-libfuzzer        │ │
> │ │       dockerfile:                                                                                          │ │
> │ │ /home/andrew/post/oss-crs-tkqdldk/.oss-crs-workdir/crs_compose/crs_src/crs-libfuzzer/oss-crs/dockerfiles/b │ │
> │ │ uilder.Dockerfile                                                                                          │ │
> │ │       args:                                                                                                │ │
> │ │         - target_base_image=sanity-mock-c-delta-01:4c6d77cdc8f0                                            │ │
> │ │         - crs_version=1.0.0                                                                                │ │
> │ │       additional_contexts:                                                                                 │ │
> │ │         libcrs: /home/andrew/post/oss-crs-tkqdldk/libCRS                                                   │ │
> │ │     cpuset: "4-129"                                                                                        │ │
> │ │     mem_limit: "16G"                                                                                       │ │
> │ │     volumes:                                                                                               │ │
> │ │       -                                                                                                    │ │
> │ │ /home/andrew/post/oss-crs-tkqdldk/.oss-crs-workdir/crs_compose/43e5a06b0d32/address/builds/1775158626g9/cr │ │
> │ │ s/crs-libfuzzer/sanity-mock-c-delta-01_4c6d77cdc8f0/BUILD_OUT_DIR:/OSS_CRS_BUILD_OUT_DIR                   │ │
> │ │ ---                                                                                                        │ │
> │ ╰────────────────────────────────────────────────────────────────────────────────────────────────────────────╯ │
> │ ╭─────────────────────────────────────────────── Error: build ───────────────────────────────────────────────╮ │
> │ │ Command failed with exit code 1:                                                                           │ │
> │ │ docker compose -p crs_582rjhwn5c -f /tmp/da3y01n0ad/docker-compose.yaml run --rm target_builder            │ │
> │ │                                                                                                            │ │
> │ │  Network crs_582rjhwn5c_default Creating                                                                   │ │
> │ │  Network crs_582rjhwn5c_default Created                                                                    │ │
> │ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Creating                                         │ │
> │ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Error response from daemon: Requested CPUs are   │ │
> │ │ not available - requested 4-129, available: 0-127                                                          │ │
> │ │ Error response from daemon: Requested CPUs are not available - requested 4-129, available: 0-127           │ │
> │ │ 📝 Docker compose output:                                                                                  │ │
> │ │ ---                                                                                                        │ │
> │ │ Command failed with exit code 1:                                                                           │ │
> │ │ docker compose -p crs_582rjhwn5c -f /tmp/da3y01n0ad/docker-compose.yaml run --rm target_builder            │ │
> │ │                                                                                                            │ │
> │ │  Network crs_582rjhwn5c_default Creating                                                                   │ │
> │ │  Network crs_582rjhwn5c_default Created                                                                    │ │
> │ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Creating                                         │ │
> │ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Error response from daemon: Requested CPUs are   │ │
> │ │ not available - requested 4-129, available: 0-127                                                          │ │
> │ │ Error response from daemon: Requested CPUs are not available - requested 4-129, available: 0-127           │ │
> │ │ ---                                                                                                        │ │
> │ │ 📝 Docker compose file contents:                                                                           │ │
> │ │ ---                                                                                                        │ │
> │ │ services:                                                                                                  │ │
> │ │   target_builder:                                                                                          │ │
> │ │     privileged: true                                                                                       │ │
> │ │     shm_size: 2g                                                                                           │ │
> │ │     environment:                                                                                           │ │
> │ │       - HELPER=True                                                                                        │ │
> │ │       - RUN_FUZZER_MODE=interactive                                                                        │ │
> │ │       - FUZZING_ENGINE=libfuzzer                                                                           │ │
> │ │       - SANITIZER=address                                                                                  │ │
> │ │       - ARCHITECTURE=x86_64                                                                                │ │
> │ │       - PROJECT_NAME=sanity-mock-c-delta-01                                                                │ │
> │ │       - FUZZING_LANGUAGE=c                                                                                 │ │
> │ │       - OSS_CRS_RUN_ENV_TYPE=local                                                                         │ │
> │ │       - OSS_CRS_BUILD_ID=1775158626g9                                                                      │ │
> │ │       - OSS_CRS_BUILD_OUT_DIR=/OSS_CRS_BUILD_OUT_DIR                                                       │ │
> │ │       - OSS_CRS_TARGET=sanity-mock-c-delta-01                                                              │ │
> │ │       - OSS_CRS_PROJ_PATH=/OSS_CRS_PROJ_PATH                                                               │ │
> │ │       - OSS_CRS_TARGET_PROJ_DIR=/OSS_CRS_PROJ_PATH                                                         │ │
> │ │       - OSS_CRS_REPO_PATH=/src/mock-c                                                                      │ │
> │ │     build:                                                                                                 │ │
> │ │       context: /home/andrew/post/oss-crs-tkqdldk/.oss-crs-workdir/crs_compose/crs_src/crs-libfuzzer        │ │
> │ │       dockerfile:                                                                                          │ │
> │ │ /home/andrew/post/oss-crs-tkqdldk/.oss-crs-workdir/crs_compose/crs_src/crs-libfuzzer/oss-crs/dockerfiles/b │ │
> │ │ uilder.Dockerfile                                                                                          │ │
> │ │       args:                                                                                                │ │
> │ │         - target_base_image=sanity-mock-c-delta-01:4c6d77cdc8f0                                            │ │
> │ │         - crs_version=1.0.0                                                                                │ │
> │ │       additional_contexts:                                                                                 │ │
> │ │         libcrs: /home/andrew/post/oss-crs-tkqdldk/libCRS                                                   │ │
> │ │     cpuset: "4-129"                                                                                        │ │
> │ │     mem_limit: "16G"                                                                                       │ │
> │ │     volumes:                                                                                               │ │
> │ │       -                                                                                                    │ │
> │ │ /home/andrew/post/oss-crs-tkqdldk/.oss-crs-workdir/crs_compose/43e5a06b0d32/address/builds/1775158626g9/cr │ │
> │ │ s/crs-libfuzzer/sanity-mock-c-delta-01_4c6d77cdc8f0/BUILD_OUT_DIR:/OSS_CRS_BUILD_OUT_DIR                   │ │
> │ │ ---                                                                                                        │ │
> │ ╰────────────────────────────────────────────────────────────────────────────────────────────────────────────╯ │
> │ ╭─────────────────────────── Error: Build target by executing the docker compose ────────────────────────────╮ │
> │ │ Command failed with exit code 1:                                                                           │ │
> │ │ docker compose -p crs_582rjhwn5c -f /tmp/da3y01n0ad/docker-compose.yaml run --rm target_builder            │ │
> │ │                                                                                                            │ │
> │ │  Network crs_582rjhwn5c_default Creating                                                                   │ │
> │ │  Network crs_582rjhwn5c_default Created                                                                    │ │
> │ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Creating                                         │ │
> │ │  Container crs_582rjhwn5c-target_builder-run-5f9d03e48a25 Error response from daemon: Requested CPUs are   │ │
> │ │ not available - requested 4-129, available: 0-127                                                          │ │
> │ │ Error response from daemon: Requested CPUs are not available - requested 4-129, available: 0-127           │ │
> │ ╰────────────────────────────────────────────────────────────────────────────────────────────────────────────╯ │
> ╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
> ```

Hi, sorry for the late response, so the validation only runs before run currently, as I understood that would be the only need. I'll extend the check to build-target so the warning shows here too.

**Comment by azchin** [comment]
Hi, sorry for the delayed feedback as we were working on a big internal refactor. I'll take a proper look this weekend, in the meanwhile could you check the CI failure? In `main` branch we also now have a `uv run verify` command (which calls https://github.com/ossf/oss-crs/blob/main/scripts/verify.py) that mimics the "CI / Verify" workflow, so you can verify locally.

**Comment by tkqdldk** [comment]
No worries ^^ Resolved the CI failure on my part and added the check in the build-target step. Let me know if I need to add or adjust anything once you take a look.

File: oss_crs/src/crs_compose.py [hunk 6]
```diff
@@ -653,7 +661,12 @@ def build_target(
                 return False
             resolved_source_path = source_dir
 
-        tasks = []
+        tasks: list[tuple[str, Callable[[MultiTaskProgress], TaskResult]]] = [
```
Comment (azchin) [inline oss_crs/src/crs_compose.py]: Could you call `_validate_machine_resources` near the start of `build()`, before the `target.build_docker_image()` gets run? Don't need to adhere to the `tasks` loop if calling it directly makes it easier.

File: oss_crs/src/crs_compose.py [hunk 7]
```diff
@@ -918,6 +931,51 @@ def _validate_required_inputs(
...
+            try:
+                total_memory_required += parse_memory_to_bytes(cfg.memory)
+            except ValueError as e:
+                log_warning(f"Failed to parse memory for {name}: {e}")
+
+        cpu_check = max_cpus_required < machine_cpu_count
+        memory_check = total_memory_required <= machine_memory
+
+        if not cpu_check or not memory_check:
+            log_warning(
```
Comment (azchin) [inline oss_crs/src/crs_compose.py]: Let's produce an error instead of warning, and terminate early. Current behavior is I see the warning message, but the docker compose still tries to run.

**Review by azchin — CHANGES_REQUESTED** [review]
I'd like the check to be placed earlier, and error + terminate.

**Comment by azchin** [comment]
CI failed b/c of LLM API key not being passed correctly, I'll look into that.

**Comment by tkqdldk** [comment]
Hi, sorry for the delay, I was very busy lately. Added the changes according to your comments.

**Review by azchin — APPROVED** [review]

**Comment by azchin** [comment]
Looks good to me! There's an integration test failure because it looks like we're overallocating memory somehow.

- Please double check whether the memory parsing is correct, since I thought we allocated 16GB for the CI machine but the warning in the integration test says "15G"
- If it's a simple rounding error, then I think it's fine to underestimate our available memory. In which case just apply the following patch to decrease the memory allocated for the builder sidecar integration test CRSs

```diff
+++ b/oss_crs/tests/integration/test_builder_sidecar_full.py
@@ -33,11 +33,11 @@ def sidecar_full_compose(tmp_dir):
     content = {
         "run_env": "local",
         "docker_registry": "local",
-        "oss_crs_infra": {"cpuset": "0-3", "memory": "8G"},
+        "oss_crs_infra": {"cpuset": "0-3", "memory": "7G"},
         "builder-sidecar-full": {
             "source": {"local_path": str(CRS_FIXTURE)},
             "cpuset": "0-3",
-            "memory": "8G",
+            "memory": "7G",
         },
     }
     path = tmp_dir / "compose.yaml"
diff --git a/oss_crs/tests/integration/test_builder_sidecar_lite.py b/oss_crs/tests/integration/test_builder_sidecar_lite.py
index 9c71c17..237923b 100644
--- a/oss_crs/tests/integration/test_builder_sidecar_lite.py
+++ b/oss_crs/tests/integration/test_builder_sidecar_lite.py
@@ -36,11 +36,11 @@ def sidecar_lite_compose(tmp_dir):
     content = {
         "run_env": "local",
         "docker_registry": "local",
-        "oss_crs_infra": {"cpuset": "0-3", "memory": "8G"},
+        "oss_crs_infra": {"cpuset": "0-3", "memory": "7G"},
         "builder-sidecar-lite": {
             "source": {"local_path": str(CRS_FIXTURE)},
             "cpuset": "0-3",
-            "memory": "8G",
+            "memory": "7G",
         },
     }
     path = tmp_dir / "compose.yaml"
```

**Comment by azchin** [comment]
Hello @tkqdldk , are you going to continue working on this PR? If you no longer have interest or bandwidth, I'll close the PR and integrate your contributions in another branch + PR so that I can make some modifications.

**Comment by tkqdldk** [comment]
> Hello @tkqdldk , are you going to continue working on this PR? If you no longer have interest or bandwidth, I'll close the PR and integrate your contributions in another branch + PR so that I can make some modifications.

Hi, sorry again for the delay. It seemed to be a rounding error from integer division the last time I checked. I can still fix it if necessary or if you choose so.

**Comment by azchin** [comment]
You can go ahead and fix it. Just wanted to check up on this

**Comment by azchin** [comment]
Could you check the pyright error? I'll merge as soon as the CI passes, your code looks good.

**Comment by tkqdldk** [comment]
Hi, I looked through the pyright error outputs and none seem related to the resource check nor the error/termination which follow the failing of the check.

**Comment by azchin** [comment]
Weird, upstream doesn't have pyright errors. I'll merge main into your branch

File: oss_crs/src/crs_compose.py [hunk 8]
```diff
@@ -63,6 +67,11 @@ def __init__(
             )
             for name, crs_cfg in self.config.crs_entries.items()
         ]
+
+        builder_count = sum(1 for crs in self.crs_list if crs.config.is_builder)
```
Comment (azchin) [inline oss_crs/src/crs_compose.py:71]: I think pyright is complaining about this line? `is_builder`
